Doc/tools/extensions/grammar_snippet.py

```diff
@@ -7,21 +12,38 @@
 from sphinx.util.docutils import SphinxDirective
 from sphinx.util.nodes import make_id
 
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+    from typing import Any
+
+    from docutils.nodes import Node
+    from sphinx.application import Sphinx
+    from sphinx.util.typing import ExtensionMetadata
 
-class SnippetStringNode(nodes.inline):
+
+class snippet_string_node(nodes.inline):  # noqa: N801 (snake_case is fine)
     """Node for a string literal in a grammar snippet."""
 
-    def __init__(self) -> None:
+    def __init__(
+        self,
+        rawsource: str = '',
+        text: str = '',
+        *children: Node,
+        **attributes: Any,
+    ) -> None:
+        super().__init__(rawsource, text, *children, **attributes)
         # Use the Pygments highlight class for `Literal.String.Other`
-        super().__init__(classes=['sx'])
+        self['classes'].append('sx')
 
 
 class GrammarSnippetBase(SphinxDirective):
     """Common functionality for GrammarSnippetDirective & CompatProductionList."""
 
     # The option/argument handling is left to the individual classes.
 
-    def make_grammar_snippet(self, options, content):
+    def make_grammar_snippet(
+        self, options: dict[str, Any], content: Sequence[str]
```

Comment:
Currently `options` is `str -> str` as the only key is `group`, but future options may have different allowed values.